File: NodeJS/src/DataStructures/collections/LinkedListDouble.js
export { DoublyLinkedList };

export class DllNode {
  constructor(_value = 0, _next = null, _prev = null) {
    this.value = _value;
    this.next = _next;
    this.prev = _prev;
  }
}

// Currently, minimal bound checking and error handling. assumes good input/testing
class DoublyLinkedList {
  constructor() {
    this.head = null;
    this.tail = null;
    this.length = 0;
  }

  /* Access */
  get(index) {
    if (this.head !== null) {
      let node = this.head;
      while (index > 0) {
        node = node.next;
        --index;
      }
      return node;
    }
    return undefined;
  }

  getBack(index) {
    if (this.head !== null) {
      let node = this.tail;
      while (index > 0) {
        node = node.prev;
        --index;
      }
      return node;
    }
    return undefined;
  }

  /* Updates */
  set(index, value) {
    let node = this.get(index);
    if (!node) return false;

    node.value = value;

    return true;
  }

  /* Insertions */
  prepend(value) {
    //O(1)
    let node = new DllNode(value, this.head);

    if (!this.head) this.tail = node;
    this.head = node;

    ++this.length;
  }

  append(value) {
    //O(1)
    let node = new DllNode(value);

    if (!this.head) {
      // Edge Case: We currently have no elements in the list.
      this.head = node;
      this.tail = node;
    } else {
      this.tail.next = node;
      node.prev = this.tail;
      this.tail = node;
    }

    ++this.length;
  }

  insert(index, value) {
    if (index < 0 || index > this.length) return false;

    if (index === this.length) this.append(value);
    else if (index === 0) this.prepend(value);
    else {
      let nodeAt = this.get(index);
      let node = new DllNode(value, nodeAt, nodeAt.prev);
      nodeAt.prev.next = node;
      nodeAt.prev = node;

      ++this.length;
    }
    return true;
  }

  /* Deletions */
  popFront() {
    //O(1)
    if (!this.head) return undefined;

    let node = this.head;
    this.head = node.next;
    this.head.prev = null;

    --this.length;

    //Edge Case: emptied list
    if (!this.head) this.tail = null;

    return node;
  }

  pop() {
    //O(n)
    if (!this.head) return undefined;

    let node = this.tail;

    if (this.length > 1) {
      node.prev.next = null;
      this.tail = this.tail.prev;
    } else {
      //Edge Case: 1 item in list
      this.head = null;
      this.tail = null;
    }

    --this.length;

    return node;
  }

  remove(index) {
    if (index < 0 || index >= this.length) return false;

    if (index === 0) this.popFront();
    else if (index === this.length - 1) this.pop();
    else {
      let nodeAt = this.get(index);

      nodeAt.prev.next = nodeAt.next;
      nodeAt.next.prev = nodeAt.prev;

      --this.length;
    }

    return true;
  }

  /* Utility */
  isEmpty() {
    return this.head === null;
  }

  /* Iterators */
  [Symbol.iterator]() {
    let node = this.head;
    return {
      next: () => {
        if (!node) return { done: true };

        const value = node.value;
        node = node.next;
        return {
          done: false,
          value,
        };
      },
    };
  }
}

File: Python/src/dss/collections/linkedlist_double.py

import colorama


class DllNode:
  def __init__(self, _val=0, _next=None, _prev=None):
    self.value = _val
    self.next = _next
    self.prev = _prev

  def __str__(self) -> str:
    return f"Value: {self.value} | Prev == None: {self.prev == None} | Next === None: {self.next == None}"


class DoublyLinkedList:
  def __init__(self):
    self.head = None
    self.tail = None
    self.__size = 0

  def __len__(self):
    return self.__size

  def __str__(self):
    curr = self.head
    node_values = []
    while curr:
      node_values.append(colorama.Fore.GREEN + str(curr.value))
      node_values.append(colorama.Style.RESET_ALL + "<->")
      curr = curr.next
    return "".join(node_values[:len(node_values)-1])

  ### Access ###
  def get(self, index):
    if index < 0 or index >= self.__size:
      raise IndexError("Index out of range")

    curr = self.head
    while index > 0:
      curr = curr.next
      index -= 1

    return curr

  ### Insertions ###
  def append(self, value):
    node = DllNode(value, None, self.tail)
    if not self.head:
      # no elements in list
      self.tail = node
      self.head = self.tail
    else:
      self.tail.next = node
      self.tail = self.tail.next

    self.__size += 1
    return True

  def append_front(self, value):
    node = DllNode(value, self.head)
    if not self.head:
      # no elements in list
      self.head = node
      self.tail = self.head
    else:
      self.head = node

    self.__size += 1
    return True

  def insert(self, index, value):
    if index == self.__size:
      self.append(value)
    elif index == 0:
      self.append_front(value)
    else:
      prev = self.get(index - 1)
      node = DllNode(value, prev.next, prev)

      prev.next.prev = node
      prev.next = node

      self.__size += 1
    return True

    ### Deletions ###

  ### Deletions ###
  def pop(self):
    if not self.head:
      return None

    node = self.tail
    if self.__size == 1:
      self.head = self.tail = None
    else:
      self.tail = self.tail.prev
      self.tail.next = None

    self.__size -= 1
    return node

  def pop_front(self):
    if not self.head:
      return None

    node = self.head
    if self.__size == 1:
      self.head = self.tail = None
    else:
      self.head = node.next
      self.head.prev = None

    self.__size -= 1

    return node

  def remove(self, index):
    if index == self.__size:
      return self.pop()
    if index == 0:
      return self.pop_front()

    node = self.get(index)

    node.prev.next = node.next
    node.next.prev = node.prev

    self.__size -= 1

    return node

  ### Utils ###

  ### Private Functions ###
